### PR TITLE
Process interrupts before reserving VMEM

### PR DESCRIPTION
**Pull request by foyzur**
Traditionally we depend on voluntary CHECK_FOR_INTERRUPTS() call to process interrupts. This hinders timely enforcement of query cancellation during critical system resource consumption such as additional memory allocation. This PR is calling ProcessInterrupts() before reserving more VMEM to ensure that we don't gobble up unreasonable amount of memory from the system before checking to see if there is a cancellation request pending.

### CODE REVIEW

**Comment by ivannovick** [comment]
I can't speak to implementation safety, but sounds like a great improvement !!!

File: src/backend/utils/mmgr/vmem_tracker.c [hunk 1]
```diff
@@ -494,7 +494,29 @@ VmemTracker_ReserveVmem(int64 newlyRequestedBytes)
...
 		RedZoneHandler_DetectRunawaySession();
+
+		/*
+		 * Before reserving further VMEM, check if the current session has a pending
+		 * query cancellation or other pending interrupts. This ensures more responsive
+		 * interrupt processing, including query cancellation requests without depending
+		 * on CHECK_FOR_INTERRUPTS(). In a sense, this is a lightweight CHECK_FOR_INTERRUPTS
+		 * as we don't execute BackoffBackendTick() and some runaway detection code.
+		 */
+		if (vmem_process_interrupt && InterruptPending)
```
Comment (ashwinstar) [inline src/backend/utils/mmgr/vmem_tracker.c]: Why GUC controlled and that too with default as false ?

Comment (vraghavan78) [inline src/backend/utils/mmgr/vmem_tracker.c]: We have a customer who is partnering with us to test in their dev system for testing this feature. We are working with them and our internal team to test this approach.

So until this is done, we cannot turn it "on" by default. For the customer to get it, we need to add the knob to turn it on/off as needed.

**Comment by vraghavan78** [comment]
LGTM

File: src/backend/utils/mmgr/vmem_tracker.c [hunk 2]
```diff
@@ -494,7 +494,29 @@ VmemTracker_ReserveVmem(int64 newlyRequestedBytes)
...
+		{
+			/* ProcessInterrupts should check for InterruptHoldoffCount and CritSectionCount */
+			ProcessInterrupts();
```
Comment (asimrp) [inline src/backend/utils/mmgr/vmem_tracker.c]: Isn't `VmemTracker_ReserveVmemChunks()` a better place for checking for interrupts?  I am not familiar with this part of the code and haven't looked at the definition of `VmemTracker_ReserveVmemChunks()` so forgive me if the suggestion is absurd!

Comment (foyzur) [inline src/backend/utils/mmgr/vmem_tracker.c]: VmemTracker_ReserveVmemChunks() is a special method that composes the broader reservation process, and that only does the actual chunk increment business. Yes, we can locate ProcessInterrupts there. But the basic idea is to separate all the peripheral things from the counter increment.

File: src/backend/utils/mmgr/vmem_tracker.c (hunk 1 above)
Comment (ashwinstar) [inline src/backend/utils/mmgr/vmem_tracker.c]: So we do not know of way to exercise this code ? There is no way for us validate the change. Still trying to understand what confirmation we wish seek. Either it works or doesn't, right. Are there some more variables involved.

**Pull request merged**